### DISCUSSION TITLE
How to parse a possibly null json string  in C++! need  help!!!!!!!!!

### DISCUSSION THREAD

**Discussion by aijyo** [nlohmann/json#4240] [Q&A] [answered]
Just as the title, server  give me a JSON string  maybe a value like:
 {"label": null}    

My C++ struct :
class LabelItem
{
public：
std::string label;
// etc ...

NLOHMANN_DEFINE_TYPE_INTRUSIVE(LabelItem, label);
}

when I parse data from server：
LabelItem item;
from_json(message, item);

My program crashed!
I've tried wrapping my string to handle null, but that failed。code is:
	class JsonString 
	{
	public:
		JsonString(const std::string& data)
			: value(data)
		{

		}
		JsonString& operator=(const std::string& val) 
		{
			this->value = val;
			return *this;
		}

		friend void to_json(nlohmann::json& j, const JsonString& data)
		{
			j = data.value;
		}

		friend void from_json(const nlohmann::json& j, JsonString& data)
		{
			if (!j.is_null())
			{
				data.value = j;
			}
		}

		std::string value;
	};

error is :
1>xxx-proto.h(189,1): error C2672: “nlohmann::json_abi_v3_11_2::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::json_abi_v3_11_2::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>>::get_to”: 未找到匹配的重载函数
1>xxx-proto.h(189,1): error C2783: “ValueType &nlohmann::json_abi_v3_11_2::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::json_abi_v3_11_2::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>>::get_to(ValueType &) noexcept(<expr>) const”: 未能为“__formal”推导 模板 参数

 Can anyone help me?

**Comment by gregmarr** [comment]
`NLOHMANN_DEFINE_TYPE_INTRUSIVE` creates functions that requires that all the values are there in the JSON and have the proper types.  If you want it to be able to use a default value when the value isn't in the JSON, you can use `NLOHMANN_DEFINE_TYPE_INTRUSIVE_WITH_DEFAULT` instead.  However, I'm pretty sure that won't handle the `null` case, and you'll need to write your own `from_json` function for that.

**Reply by aijyo** [reply]
After my  define my own type to process null string， nested struct cannot pass compile：
JsonString is:
	class JsonString 
	{
	public:
		JsonString(const std::string& data)
			: value(data)
		{

		}
		JsonString& operator=(const std::string& val) 
		{
			this->value = val;
			return *this;
		}

		friend void to_json(nlohmann::json& j, const JsonString& data)
		{
			j = data.value;
		}

		friend void from_json(const nlohmann::json& j, JsonString& data)
		{
			if (!j.is_null())
			{
				data.value = j;
			}
		}

		std::string value;
	};

Nested class:
	class SubscribeRequest
	{
	public:
		class SubPublisherItem
		{
		public:
			class SubOptions
			{
			public:
				class SubSimulcast
				{
				public:
					JsonString profile;

				public:
					NLOHMANN_DEFINE_TYPE_INTRUSIVE(SubSimulcast, profile);
				};

			public:
				SubSimulcast simulcast;

			public:
				NLOHMANN_DEFINE_TYPE_INTRUSIVE(SubOptions, simulcast);
			};
		public:
			JsonString sessionId;
			JsonString userId;
			SubOptions options;

		public:
			NLOHMANN_DEFINE_TYPE_INTRUSIVE(SubPublisherItem, sessionId, userId);
		};

	public:
		JsonString event = SUBSCRIBE_MSG;
		JsonString roomToken;
		JsonString userId;
		JsonString subscriberType;
		JsonString txid;
		std::vector<SubPublisherItem> publishers;
	public:
		NLOHMANN_DEFINE_TYPE_INTRUSIVE(SubscribeRequest, event, roomToken, userId
			, subscriberType, publishers, txid);
	};

**error**:
严重性	代码	说明	项目	文件	行	禁止显示状态
错误	C2672	“nlohmann::json_abi_v3_11_2::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::json_abi_v3_11_2::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>>::get_to”: 未找到匹配的重载函数	xxx	xxx-proto.h	417	
错误	C2783	“ValueType &nlohmann::json_abi_v3_11_2::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::json_abi_v3_11_2::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>>::get_to(ValueType &) noexcept(<expr>) const”: 未能为“__formal”推导 模板 参数   xxx	xxx-proto.h	417

**Reply by gregmarr** [reply, marked as answer]
I'm not sure why it doesn't pick up the `from_json` in the `JsonString` for the `SubPublisherItem`, but I wasn't suggesting you create a wrapper type for `std::string`.  I was suggesting that you implement the from_json in the container classes, as the method you're trying won't handle a missing value.  With the latest version you can use `NLOHMANN_DEFINE_TYPE_INTRUSIVE_ONLY_SERIALIZE` to do the `to_json` part.

Having said that, using `NLOHMANN_DEFINE_TYPE_INTRUSIVE_WITH_DEFAULT` in the containing classes works once you add a default constructor to `JsonString`.  https://www.godbolt.org/z/W61P8jfeE

**Reply by aijyo** [reply]
Thanks for your help，I decided to implement from_json myself, but I think it is also necessary to leave a method that can convert null to empty string, because there are really many places where this string needs to be used, and I cannot ask the server to modify the code. Others may also want this situation。
